enhancements/cloud-integration/tokenized-auth-enablement-operators-on-cloud.md

```diff
@@ -0,0 +1,404 @@
+as part of this EP)
+- Add "role-arn"-type fields to the CredentialRequests
+- Add eventing to report status on a CR to indicate lacking STS credentials for fully operational deploy or update.
+
+For the Cloud Credential Operator:
+- Allow detection that a cluster is using time-based tokens and create a Secret containing the ARN + token path per 
+ cloud platform.
+
+For Operator Administrator:
+- Supply cloud credentials for STS (ARN ID, etc) in the known location.
+
+For OperatorHub, Console:
+- Prompt for missing credentials that prevent fully operational install on STS cluster. Same for upgrades.
+- Badge indicating "STS enabled" (this is a naming suggestion to avoid implying that non-badged operators not needing
+ cloud provider resource access would not work) operators.
+
+This graph shows the system as it needs to work (AWS only), not the proposed solution workflow:
+```mermaid
+graph  LR
+  sa([ServiceAccount Signing Keys]) -- Public <br> Key -->s3[S3 Bucket w/ OIDC Config];
+  sa([ServiceAccount Signing Keys]) -- Private <br> Key -->token_signing;
+  service<-.->config
+  service<-.->Pod
+  Pod-->role
+  subgraph cluster [AWS]
+  s3;
+  iam_id[IAM Identity Provider] -- Trust Tokens<br>Signed -->s3;
+  role[IAM Role For Pod]<-->service[AWS Services];
+  end
+  subgraph cluster2 [OpenShift]
+  token_signing[Token Signing]-->projected[Projected<br>ServiceAccount<br>Token];
+  subgraph Pod
+  config[AWS Config File]-->projected
+  end
+  end
+  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
+  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
+  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
+  classDef cluster2 fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
+  class config,role,service,s3,projected,token_signing,iam_id,pod1,pod2 k8s;
+  class sa plain;
+  class cluster cluster;
+  class cluster2 cluster1;
+```
+
+### API Extensions
+
+Added to the CredentialsRequest API 
+
+We outline this change using the AWS specific change, but other cloud provider specs will follow a similar pattern:
+Adding to the `spec.ProviderSpec` field ([ref](https://github.com/openshift/cloud-credential-operator/blob/38e7f96bf73036924abcb1352222c17704e4b2e0/pkg/apis/cloudcredential/v1/types_credentialsrequest.go#L53-L55)),<br>
+where `spec.ProviderSpec` would be an `AWSProviderSpec` ([ref](https://github.com/openshift/cloud-credential-operator/blob/38e7f96bf73036924abcb1352222c17704e4b2e0/pkg/apis/cloudcredential/v1/types_aws.go#L27-L31)) for AWS-directed TAT.<p>
+CredentialsRequests:
+```go
+type CredentialsRequestSpec struct {
+...
+
+// ProviderSpec contains the cloud provider specific credentials specification.
+// +kubebuilder:pruning:PreserveUnknownFields
+ProviderSpec *runtime.RawExtension `json:"providerSpec,omitempty"`
+
+...
+// CloudTokenPath (JWT token)
+// +optional
+CloudTokenPath string `json:"cloudTokenPath,omitempty"`
```

Comment:
This needs more context, I've read through the enhancement and I'm not entirely sure what this is for. Is this the path that the operator expects the token to be available at within the pod?

Imagine I'm someone who hasn't read the enhancement, explain in the godoc everything I would need to know

Reply:
yeah, not clear to me what component consumes this value or what they would do with it.

Reply:
The AWS STS credentials secret looks like this and `CloudTokenPath` ultimately ends up in the secret CCO generates as the `web_identity_token_file` the AWS SDK expects.

```
apiVersion: v1
stringData:
  credentials: |-
    [default]
    role_arn = <role ARN>
    web_identity_token_file = <path to mounted service account token such /var/run/secrets/openshift/serviceaccount/token>
kind: Secret
type: Opaque
```

Similarly, the Azure credentials secret CCO generates contains an `azure_federated_token_file` field.

```
apiVersion: v1
data:
  azure_client_id: <client id>
  azure_federated_token_file: <path to mounted service account token such as /var/run/secrets/openshift/serviceaccount/token>
  azure_region: <region>
  azure_subscription_id: <subscription id>
  azure_tenant_id: <tenant id>
kind: Secret
type: Opaque
```

It's the path where the service account token was mounted on the deployment.

Here's a pass at a comment descriptor,

```
// cloudTokenPath is the path where the Kubernetes ServiceAccount token (JSON Web Token) is mounted
// on the deployment for the workload requesting a credentials secret.
// The presence of this field in combination with fields such as spec.providerSpec.stsIAMRoleARN
// indicate that CCO should broker creation of a credentials secret containing fields necessary for
// token based authentication methods such as with the AWS Secure Token Service (STS).
// +optional
CloudTokenPath string `json:"cloudTokenPath,omitempty"`
```

Reply:
so the short version is that the value of this field gets injected into the Secret that the CCO creates.

the slightly longer version is that the serviceaccounttoken needs to be injected/mounted into that pod at the same location, but there is no central component that is going to ensure that is the case/validate that the credreq value and the mount/injection path are aligned.

basically it's up to the operator to define the credreq value correctly, especially since the operator itself doesn't even control the mount, k8s controls/chooses the SAToken mount path.

which makes me wonder if it should even be a configurable field value?  When would this field take on any value other than `/var/run/secrets/kubernetes.io/serviceaccount`

I guess i can envision a future state where the tokenpath might be something else, but at least for our immediate use case i think 100% of the time that will be the value, right?

Reply:
I am integrating Andrews expansions. We can resolve this thread.